File: leetcode/src/Num121.java
public class Num121 {
    class Solution {
    /**
     * 思路：这题可以用动态规划去解决问题
     * 令dp[i] 表示第i天售出时的最大收益，那么dp[i] = max(dp[i-1] + price[i] - price[-1],price[i] - price[-1])
     * 其中dp[0] = 0*/
    public int maxProfit(int[] prices) {
        int max = 0;
        if(prices.length < 2){return 0;}
        int [] dp = new int [prices.length];
        dp[0] = 0;
        for (int i = 1; i < dp.length; i++){
            dp[i] = prices[i] - prices[i - 1];
            if(dp[i - 1] > 0){
                dp[i] += dp[i - 1];
            }
            if(max < dp[i]){max = dp[i];}
        }
        return max;
    }
}
}

File: learnPython/num121.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        sz = len(prices)
        if sz <= 1:
            return 0
        imax = 0
        dp = list(range(sz))
        for i in range(1,sz):
            dp[i] = max(prices[i]-prices[i-1] + dp[i-1],prices[i]-prices[i-1])
            if dp[i] > imax:
                imax = dp[i]
        return imax
if __name__=='__main__':
    s = Solution()
    input = [7,1,5,3,6,4]
    print(s.maxProfit(input))
    input = [7,6,4,3,1]
    print(s.maxProfit(input))